protocol/strict-derivation.md

```diff
@@ -0,0 +1,269 @@
+can be derived from a span batch are valid, or if any of them is invalid (during batch queue
+checks), the whole span batch is dropped. This has negative implications for Fault Proofs: in a
+worst-case scenario, if the to-be-fault-proven block is the first batch in a span batch, to
+determine its validity at the batch queue stage, the validity of the full span batch has to be
+determined, as any later invalid singular batch would also invalidate all prior batches in the same
+span batch. So only invalidating the remaining span batch upon hitting an invalid derived singular
+batch vastly improves this scenario, since batch queue validity is then final for any singular
+batch, not just regular singular batches.
+
+In late 2023 while Delta was still in development, the problem of partial span batch validity was
+[already discussed](https://www.notion.so/oplabs/Handling-Invalid-Batches-48bb972368ce409c84bd9189eddd0577)
+and for simplicity it was decided to use the atomic validity model. However, above discussion
+focusses on implications of the atomic validity model for EL-sync. We're now concerned with behavior
+around CL client restarts and reorgs.
+
+### Mid-way restart
+
+One concern that was raised with partial span batch validity is how to deal with mid-way restarts,
+i.e. when the CL client restarts, loosing its internal derivation pipeline state while the safe head
+of the EL client is in the middle of a span batch. In particular, the correct behavior needs to
+cover the case where a L1 reorg happened in the meantime that reorged out the L1 block that
+contained the span batch.
+
+I fail to see currently how partial span batch validity makes this case worse or different. With
+atomic span batch validity, there's the pending-safe state of derived blocks that might be reorged
+out later by a backwards-invalidating singular batch derived from a span batch. So if we only
+forwards-invalidate, this becomes easier.
+
+And in any case, a restart or reorg needs to reconstruct the derivation pipeline state from the new
+(sync-start) L1 block, so I don't understand how reconstructing span batch derivation state is
+different from reconstructing the channel bank and batch queue in general. If a restart or reorg
+happens in the middle of deriving singular batches from a channel, we're in a similar situation that
+is solved by going back far enough for the sync-start.
+
+## Proposed Design
+
+To resolve open questions regarding the implementation of Partial Span Batch Validity, I propose to
+apply the principle to consistently change the validation rules to be only forward-invalidating, and
+thus treat a span batch like an additional (optional) stage in the derivation pipeline that sits
+before the batch queue, so that the batch queue pulls singular batches from this previous span batch
+stage. This way, we never backwards-invalidate the span batch, only forward.
```

Comment:
Curious if the primary driver of this is to remain backwards compatible. It seems to me that this new stage would be very closely similar to the role of the batch queue itself as noted in [the inline docs](https://github.com/ethereum-optimism/optimism/blob/3f7a8ad1b8030a3d01d44bc57bf0553a249eae23/op-node/rollup/derive/batch_queue.go#L15-L28). I guess my question here would then be - is it worthwhile introducing a new stage as opposed to modifying the batch queue?

Reply:
I think an _implementation_ can choose to do either, have it as a separate stage, or part of the batch queue. I like the modularity of thinking of it as a separate stage _in the abstract_.